src/game-servers/providers/serveme-tf/services/serveme-tf-api.service.ts

```diff
@@ -0,0 +1,223 @@
+  private readonly config: AxiosRequestConfig = {
+    params: {
+      api_key: this.environment.servemeTfApiKey,
+    },
+  };
+
+  constructor(
+    private httpService: HttpService,
+    private environment: Environment,
+  ) {}
+
+  async reserveServer(): Promise<ServemeTfReservationDetailsResponse> {
+    // https://github.com/Arie/serveme#api
+    return await lastValueFrom(
+      this.fetchServers().pipe(
+        switchMap((response) =>
+          this.httpService.post<ServemeTfReservationDetailsResponse>(
+            response.actions.create,
+            {
+              reservation: {
+                starts_at: response.reservation.starts_at,
+                ends_at: response.reservation.ends_at,
+                rcon: response.reservation.rcon,
+                password: 'test',
+                server_id: sample(
+                  response.servers
+                    // make sure we dont' take a SDR server
+                    // https://partner.steamgames.com/doc/features/multiplayer/steamdatagramrelay
+                    .filter((server) => server.sdr === false),
+                ).id,
+              },
+            },
+            this.config,
+          ),
+        ),
+        map((response) => response.data),
+        tap((reservation) =>
+          this.logger.log(
+            `serveme.tf reservation ${reservation.reservation.id} (${reservation.reservation.server.name}) created`,
+          ),
+        ),
+      ),
+    );
+  }
+
+  async waitForServerToStart(reservationId: number): Promise<void> {
+    return await lastValueFrom(
+      timer(1000, 1000).pipe(
+        exhaustMap(() => this.fetchReservationDetails(reservationId)),
+        map((reservation) => reservation.reservation.status),
+        takeWhile((status: ReservationStatus) =>
+          [
+            'Waiting to start',
+            'Starting',
+            'Server updating, please be patient',
+          ].includes(status),
+        ),
+        map(() => null),
+      ),
+    );
+  }
+
+  async endServerReservation(
+    reservationId: number,
+  ): Promise<ServemeTfReservationDetailsResponse> {
+    return await lastValueFrom(
+      this.fetchReservationDetails(reservationId).pipe(
+        switchMap((reservation) => {
+          if (!reservation.reservation.ended) {
+            return this.httpService
+              .delete<ServemeTfReservationDetailsResponse>(
+                reservation.actions.delete,
+                this.config,
+              )
+              .pipe(map((response) => response.data));
+          } else {
+            return of(reservation);
+          }
+        }),
+        tap((reservation) =>
+          this.logger.log(
+            `serveme.tf reservation ${reservation.reservation.id} (${reservation.reservation.server.name}) ended`,
+          ),
+        ),
+      ),
+    );
+  }
+
+  private fetchServers(): Observable<ServemeTfFindServersResponse> {
+    return of(`${servemeTfApiEndpoint}/new`).pipe(
```

Comment:
Why are you taking an observable of a string then piping it into an HTTP request?

Why not take from the HTTP observable instead?

Reply:
Just coding style - it's easier to follow all the `switchMap`s if they are in the same indentation block.

Reply:
IMHO I prefer to _not_ do this as it's quite redundent.